woocommerce/woocommerce
# E2e/e2e merchant orders add order complex

### All Submissions:






### Changes proposed in this Pull Request:


This Pull Request will add an e2e test that would cover the `Merchant | Orders | Add new Order complex (multiple product types & tax classes)` critical flow.
Closes #29003.
Below are the steps that this test does:

#### Test Data
1. **Products** - this test will create and use 4 products as test data:
    - One simple product named `Simple Product 273722`
    - One variable product named `Variable Product 024611`
    - One grouped product named `Grouped Product 858012`
    - One external product named `External product 786794`

    Since the test suite creates several products having the same names specified in `default.json`, I made this test's product names unique by suffixing them with a 6-digit random number, like `Simple Product 273722` so that the test would be able to select them correctly.

    I also did not follow the default `regularPrice` in `default.json` but instead specified my own, so that this test wouldn't fail when `default.json` is later modified.

2. **Tax Classes** - this test will create and use 3 tax classes:
    - One tax class for the simple product
    - One tax class for the variable product
    - One tax class for the external product

    I'm hesitant to use the default tax classes (Standard, Zero Rate, Reduced Rate, etc.) because I'm afraid that any changes to them by other tests in the future will break this test.

3. **Tax Rates** - this test will create and use 3 tax rates, one for each of the tax classes mentioned above.

#### Pre-test Setup (`beforeAll` block)
1. Using the API:
    1. Toggle on the "Enable tax rates and calculations" setting. Even though the `activate-and-setup` tests will turn this setting on, I still specified it here to be sure.
    2. Create the tax classes, tax rates, and products I mentioned above in the _"Test Data"_ section.
2. Using the UI, login as the merchant.
    
#### Test steps
1. Go to "add order" page.
2. Click 'Add item(s)' button.
3. Click 'Add Product(s)' button
4. In the modal that appears, type search for the following products and click on them in the results menu:
    - Simple Product 273722
    - Variable Product 024611
    - Grouped Product 858012
    - External product 786794
5. Click the 'Add' button in the modal window
6. Click 'Recalculate' button.
7. Click 'Create' button.
8. Verify that the table shows all the 4 product names.
9. Verify that the table headers show all the names of the 4 tax classes.
10. Verify that the table shows the expected tax rates.
11. Verify that the order totals below the table contain the correct tax names and amounts.

At the end of the test, it would have created an order containing the following products and taxes:
<img width="514" alt="HN127xT4xs" src="https://user-images.githubusercontent.com/4509348/113540375-9e0a8780-9612-11eb-94ae-5c51ab24695a.png">

### How to test the changes in this Pull Request:

1. After running `docker:up`, run the following commands:
```
# set initial state of the test
npx wc-e2e test:e2e tests/e2e/specs/activate-and-setup/setup-onboarding.test.js

# run this test
npx wc-e2e test:e2e tests/e2e/specs/wp-admin/test-create-order.js
```
2. Wait for the test to finish. You should see a passing `can create new complex order with multiple product types & tax classes` test
<img width="349" alt="Ehl7g3opZB" src="https://user-images.githubusercontent.com/4509348/113535564-57169500-9606-11eb-9ac0-ebeff3f4349c.png">

### Other information:
